# ASR Engine Evaluation

## Goal
To evaluate our existing speech recognition (ASR) engine running on whisper.cpp against other potential ASR engines for Swahili. If I am able to find a better ASR that is already ported onto 
whisper.cpp, we should move to it. 

## Background

We use OpenAI's Whisper Small model, converted to whisper.cpp, and quantized to 5-bit, without any fine-tuning, about 181MB. You can find it here: https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-small-q5_1.bin

Three native Swahili speakers dictated their answers to Survey questions 8 - 17 into the Survey app. This generated 30 .wav files that I can use to test the various ASR engines. 

Here is the text that they dictated for each question: 

| Question | Expected answer (Swahili) | English translation |
|---|---|---|
| Q8 | Ndiyo, iliharibu majani na kupunguza mavuno. | Yes, it damaged the leaves and reduced the harvest. |
| Q9 | Naweza kukubali kupoteza hadi asilimia kumi ya mavuno. | I can accept losing up to ten percent of the harvest. |
| Q10 | Ikiwa uharibifu unafika karibu asilimia thelathini, nitabadilisha aina. | If the damage reaches about thirty percent, I will change varieties. |
| Q11 | Iwe na mavuno mengi na iwe inastahimili ukame na wadudu. | It should have a high yield and be tolerant of drought and pests. |
| Q12 | Hata tani moja kwa ekari, kama mbegu ni nzuri. | Even one ton per acre, if the seed is good. |
| Q13 | Wakati wa maua na kutengeneza punje. | During flowering and grain formation. |
| Q14 | Ninaweka ya chakula, nauza iliyobaki na kidogo ni ya mifugo. | I keep some for food, sell the rest, and a little is for livestock. |
| Q15 | Nawapa kuku na nguruwe mahindi mara mbili hadi tatu kwa wiki. | I give the chickens and pigs maize two to three times a week. |
| Q16 | Ninanunua kwa agrovet kwa sababu ninaamini ubora wake. | I buy from the agrovet because I trust its quality. |
| Q17 | Nauza kwa wafanyabiashara wa eneo kwa sababu wako karibu na hulipa haraka. | I sell to local traders because they are nearby and pay quickly. |

## Pre-processing

With some of the previous test sessions, the words that were dictated did not exactly match the expected answer. To be able to deal with that, we created a .json file that had a link to all of the 
.wav files, and the expected text for each one. For this data set, we only needed to update survey 46ec4547, Q14, which said, "Naweka gunia tano ya chakula, gunia tano nauza, na mbili ni ya mifugo.". 
If there were multiple .wav files for any question, we used the latest recording. 

The original answers contained numbers, like 10%. To make the results more consistent, I updated all of the numbers to words instead of numbers. So 10% became asilimia kumi. 
Sometimes the ASR would return numbers (which would makes sense to a human reader), but that would mess up the error rates. So I post-processed the transcription to replace any numbers with text prior to error calculation. 


In [1]:
import json

In [2]:
#define the path to the reference text
# One entry per survey recording: survey_id, question_id, expected_answer, wav_file, local_wav
reference_file = "office_testing/expected_answers.json"
reference = reference_file

# Read from a relative path
with open(reference, 'r', encoding='utf-8') as g:
    ref_data = json.load(g)

# Look up the expected answer by (survey_id, question_id), since a survey can have its own reference text
ref_lookup = {(r["survey_id"], r["question_id"]): r["expected_answer"] for r in ref_data}

print (f"Reference data: loaded {len(ref_data)} entries from {reference_file}")
for r in ref_data:
    print(f"{r['survey_id'][:8]} {r['question_id']}: {r['expected_answer']}")

Reference data: loaded 30 entries from office_testing/expected_answers.json
46ec4547 Q8: Ndiyo, iliharibu majani na kupunguza mavuno.
46ec4547 Q9: Naweza kukubali kupoteza hadi asilimia kumi ya mavuno.
46ec4547 Q10: Ikiwa uharibifu unafika karibu asilimia thelathini, nitabadilisha aina.
46ec4547 Q11: Iwe na mavuno mengi na iwe inastahimili ukame na wadudu.
46ec4547 Q12: Hata tani moja kwa ekari, kama mbegu ni nzuri.
46ec4547 Q13: Wakati wa maua na kutengeneza punje.
46ec4547 Q14: Naweka gunia tano ya chakula, gunia tano nauza, na mbili ni ya mifugo.
46ec4547 Q15: Nawapa kuku na nguruwe mahindi mara mbili hadi tatu kwa wiki.
46ec4547 Q16: Ninanunua kwa agrovet kwa sababu ninaamini ubora wake.
46ec4547 Q17: Nauza kwa wafanyabiashara wa eneo kwa sababu wako karibu na hulipa haraka.
4f629181 Q8: Ndiyo, iliharibu majani na kupunguza mavuno.
4f629181 Q9: Naweza kukubali kupoteza hadi asilimia kumi ya mavuno.
4f629181 Q10: Ikiwa uharibifu unafika karibu asilimia thelathini, nitabadilisha aina

In [3]:
# Define the GitHub URLs of the surveys we want to test today.
# Paste the normal github.com "blob" links; they are converted to raw file URLs below
survey_urls = [
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-17/exports/survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json",
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-21/exports/survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json",
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-21/exports/survey_7db59ab6-e13d-4080-bb07-3842d8b51fe4_2026-09-21_08-45-52.json",
]

from urllib.request import urlopen

# One entry per survey: file name plus the reference/answer pairs that matched.
surveys = []
for survey_url in survey_urls:
    raw_url = survey_url.replace("https://github.com/", "https://raw.githubusercontent.com/", 1).replace("/blob/", "/", 1)
    with urlopen(raw_url) as f:
        survey_data = json.load(f)

    survey_file = raw_url.rsplit("/", 1)[-1]
    print(f"\nLoaded data from {survey_file}")

    # Free text entered for the session (e.g. who tested and when), if present.
    session_free_text = survey_data.get('meta', {}).get('session_free_text', '')
    print(f"Session free text: {session_free_text}")

    # Keep only the questions that have a reference answer for this survey, in question order.
    survey_id = survey_data['survey_id']
    ref_questions = sorted((q for (sid, q) in ref_lookup if sid == survey_id), key=lambda q: int(q[1:]))
    if not ref_questions:
        print(f"No reference entries for survey {survey_id}, skipped")
        continue
    questions = [q for q in ref_questions if q in survey_data['answers']]
    missing = [q for q in ref_questions if q not in survey_data['answers']]
    if missing:
        print(f"Not in survey, skipped: {missing}")

    references = [ref_lookup[(survey_id, q)] for q in questions]
    answers = [survey_data['answers'][q]['answer'] for q in questions]
    print(f"Found {len(answers)} transcribed answers matching the reference.")
    for q, ans in zip(questions, answers):
        print(f"{q}: {ans}")

    surveys.append({"file": survey_file, "references": references, "answers": answers})


Loaded data from survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json
Session free text: This is Victor Kitoto filling in the survey on 17th September 2026
Found 10 transcribed answers matching the reference.
Q8: Ndiyo, elia ribu majani na kupunguza mawuno.
Q9: Na eza kukubali kupoteza hadia silimi ya kumi ya mawono.
Q10: Ikiva huwari defu inafika karibu asile miatela tini, intaba deleisha maa aina.
Q11: Iwe na mawuna mengi na iwe inasta emili ukame na wadulu
Q12: Atata ni moja kwaekari kama mbegunin zori.
Q13: Wakati wamaoa na kutengeneza punje
Q14: Ninaweka ya chakula na uza iliobaki na kidobu ni amifugo.
Q15: Nwaapa koko nanguruwe maindi marambili hadi tatu kua wiki.
 Nwaapa koko nanguruwe maindi marambili hadi tatu kua wiki.
Q16: Nina nunua kwa agrovet kwa sababu nina aminiu borawake.
Q17: Na uza kwa fanya bia shara wa ineu kwa sababu wa kukaribu na uli paraka.

Loaded data from survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json
Session free text:

In [4]:
import jiwer
import re
from jiwer.transforms import AbstractTransform

# Swahili number words, so "30", "30%" and "thelathini" all score the same.
_UNITS = ["", "moja", "mbili", "tatu", "nne", "tano", "sita", "saba", "nane", "tisa"]
_TENS = ["", "kumi", "ishirini", "thelathini", "arobaini", "hamsini", "sitini", "sabini", "themanini", "tisini"]

def swahili_number(n):
    """Spell out 0-999,999 in Swahili, e.g. 30 -> 'thelathini', 25 -> 'ishirini na tano'."""
    if n == 0:
        return "sifuri"
    parts = []
    if n >= 1000:
        parts.append("elfu " + swahili_number(n // 1000)); n %= 1000
    if n >= 100:
        parts.append("mia " + swahili_number(n // 100)); n %= 100
    if n >= 10:
        parts.append(_TENS[n // 10]); n %= 10
    if n:
        parts.append(_UNITS[n])
    return " na ".join(parts)

class NormalizeSwahiliNumbers(AbstractTransform):
    """Digits to Swahili words: '2-3' -> 'mbili hadi tatu', '10%' -> 'asilimia kumi'."""
    def process_string(self, s):
        num = lambda m: swahili_number(int(m))
        s = re.sub(r"\b(\d+)\s*[-–]\s*(\d+)\b", lambda m: f"{num(m[1])} hadi {num(m[2])}", s)  # ranges
        s = re.sub(r"\b(?:asilimia\s+)?(\d+)\s*%", lambda m: f"asilimia {num(m[1])}", s)       # percentages
        return re.sub(r"\b\d+\b", lambda m: num(m[0]), s)                                      # plain numbers


transform_words = jiwer.Compose ([
    jiwer.ToLowerCase(),
    NormalizeSwahiliNumbers(),
    jiwer.RemovePunctuation(),
    jiwer.RemoveMultipleSpaces(),
    jiwer.Strip(), 
    jiwer.ReduceToListOfListOfWords()
])

transform_chars = jiwer.Compose ([
    jiwer.ToLowerCase(),
    NormalizeSwahiliNumbers(),
    jiwer.RemovePunctuation(),
    jiwer.RemoveMultipleSpaces(),
    jiwer.Strip(), 
    jiwer.ReduceToListOfListOfChars()
])

def error_rates(references, answers):
    out = jiwer.process_words(references, answers, reference_transform=transform_words, hypothesis_transform=transform_words)
    out2 = jiwer.process_characters(references, answers, reference_transform=transform_chars, hypothesis_transform=transform_chars)
    return out.wer, out2.cer

# Per-survey rates.
for s in surveys:
    wer, cer = error_rates(s["references"], s["answers"])
    print(f"{s['file']}  ({len(s['answers'])} answers)")
    print(f"  WER (normalized): {wer:.3f}   CER (normalized): {cer:.3f}")

# Aggregate over every answer from every survey: total errors / total reference words (or characters).
all_references = [r for s in surveys for r in s["references"]]
all_answers = [a for s in surveys for a in s["answers"]]
wer, cer = error_rates(all_references, all_answers)
print(f"\nAggregate over {len(surveys)} surveys, {len(all_answers)} answers")
print(f"  WER (normalized): {wer:.3f}   CER (normalized): {cer:.3f}")

survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json  (10 answers)
  WER (normalized): 0.780   CER (normalized): 0.296
survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json  (10 answers)
  WER (normalized): 0.659   CER (normalized): 0.120
survey_7db59ab6-e13d-4080-bb07-3842d8b51fe4_2026-09-21_08-45-52.json  (10 answers)
  WER (normalized): 0.727   CER (normalized): 0.120

Aggregate over 3 surveys, 30 answers
  WER (normalized): 0.723   CER (normalized): 0.179


## Conclusion
### Current ASR
Gives us WER = 0.723, CER = 0.179

### Other ASR Engines
We found four ASR engines on HuggingFace that focus on Swahili. 

I ran that same script against each model, with the same 30 samples. 
| Model | Hugging Face | Focus | whisper.cpp file | WER | CER |
|---|---|---|---|---|---|
| Lingua-Connect Swahili | [Lingua-Connect/whisper-small-sw-normal](https://huggingface.co/Lingua-Connect/whisper-small-sw-normal) | Swahili baseline | `ggml-small-sw-q5_0.bin` | 0.446 | 0.094 |
| CDLI CV20 | [cdli/whisper-small-Swahili_finetuned_small_CV20](https://huggingface.co/cdli/whisper-small-Swahili_finetuned_small_CV20) | Common Voice 20 | `ggml-small-sw-cv20-q5_0.bin` | 0.273 | 0.054 |
| CDLI Kenyan non-standard | [cdli/whisper-small_finetuned_kenyan_swahili_nonstandard_speech_v1.0](https://huggingface.co/cdli/whisper-small_finetuned_kenyan_swahili_nonstandard_speech_v1.0) | Kenyan / non-standard Swahili | `ggml-small-sw-kenyan-nonstandard-q5_0.bin` | 0.236 | 0.062 |
| dmusingu CV14 | [dmusingu/WHISPER-SMALL-SWAHILI-ASR-CV-14](https://huggingface.co/dmusingu/WHISPER-SMALL-SWAHILI-ASR-CV-14) | Common Voice 14 | `ggml-small-sw-cv14-q5_0.bin` | 0.146 | 0.061 |

The "dmusingu CV14" ASR looks dramatically better than what we are currently using. CER (which is what I am mostly concerned with, since Swahili is an agglutinative language), is less than 1/2 of what we have now. 
We will plan to integrate this ASR into our survey. As we get more recorded data, I will update this test. 
